src/webgpu/web_platform/external_texture/video.spec.ts

```diff
@@ -151,26 +151,105 @@ Test that importing different video formats into an external texture and samplin
+                    vec4<f32>( 1.0, -1.0, 0.0, 1.0),
+                    vec4<f32>(-1.0, -1.0, 0.0, 1.0),
+                    vec4<f32>( 1.0,  1.0, 0.0, 1.0),
+                    vec4<f32>(-1.0, -1.0, 0.0, 1.0),
+                    vec4<f32>(-1.0,  1.0, 0.0, 1.0)
+                  );
+                  return pos[VertexIndex];
+              }
+              `,
+          }),
+          entryPoint: 'main',
+        },
+        fragment: {
+          module: t.device.createShaderModule({
+            code: `
+              [[group(0), binding(0)]] var s : sampler;
+              [[group(0), binding(1)]] var t : texture_external;
+
+              [[stage(fragment)]] fn main([[builtin(position)]] FragCoord : vec4<f32>)
+                                       -> [[location(0)]] vec4<f32> {
+                  return textureSampleLevel(t, s, FragCoord.xy / vec2<f32>(16.0, 16.0));
+              }
+              `,
+          }),
+          entryPoint: 'main',
+          targets: [
             {
-              view: colorAttachment.createView(),
-              loadValue: { r: 0.0, g: 0.0, b: 0.0, a: 1.0 },
-              storeOp: 'store',
+              format: kFormat,
             },
           ],
-        });
-        passEncoder2.setPipeline(pipeline);
-        passEncoder2.setBindGroup(0, bindGroup);
-        passEncoder2.draw(6);
-        passEncoder2.endPass();
+        },
+        primitive: { topology: 'triangle-list' },
+      });
+
+      const linearSampler = t.device.createSampler();
+
+      const externalTextureDescriptor = { source: video };
+      const externalTexture = t.device.importExternalTexture(externalTextureDescriptor);
+
+      const bindGroup = t.device.createBindGroup({
+        layout: pipeline.getBindGroupLayout(0),
+        entries: [
+          {
+            binding: 0,
+            resource: linearSampler,
+          },
+          {
+            binding: 1,
+            resource: externalTexture,
+          },
+        ],
+      });
+      const commandEncoder = t.device.createCommandEncoder();
+      const passEncoder = commandEncoder.beginRenderPass({
+        colorAttachments: [
+          {
+            view: colorAttachment.createView(),
+            loadValue: { r: 0.0, g: 0.0, b: 0.0, a: 1.0 },
+            storeOp: 'store',
+          },
+        ],
+      });
+      passEncoder.setPipeline(pipeline);
+      passEncoder.setBindGroup(0, bindGroup);
+      passEncoder.draw(6);
+      passEncoder.endPass();
 
-        t.expectGPUError('validation', () => t.device.queue.submit([commandEncoder2.finish()]));
+      // requestAnimationFrame should trigger a microtask that destroys the GPUExternalTexture,
+      // which will cause a validation error when we try to use the destroyed GPUExternalTexture
+      const ensureDestroyed = function () {
+        t.expectGPUError('validation', () => t.device.queue.submit([commandEncoder.finish()]));
       };
 
       requestAnimationFrame(ensureDestroyed);
```

Comment:
This can be `timeout(ensureDestroyed, 0)`, which is a little bit stricter. (`timeout` is a util wrapper around `setTimeout`.)

If we want to be perfectly strict, we should really test that it gets scheduled as a microtask at exactly the right time. This gets kind of in the weeds, but what we can do is issue three microtasks by taking advantage of the fact that .then() on an already-resolved promise is always scheduled as a microtask:

```ts
let externalTexture;
Promise.resolve().then(() => { /* here, externalTexture should be alive */ });
externalTexture = device.importExternalTexture(/* ... */);
Promise.resolve().then(() => { /* here, externalTexture should have been destroyed */ });
await resolveOnTimeout(0); // [1]
```

(I'm not _100%_ sure this will work, so let me know if something seems wrong.)

[1] AFAICT, nothing is strictly ensuring that the test waits for this to happen before ending. Probably it works because `await startPlayingAndWaitForVideo` ends up waiting long enough after the callback for whatever reason.
With the changes above to make everything happen in microtasks, it should be enough to wait for a (non-micro) task like `setTimeout`, as microtasks always happen before tasks.

Reply:
_Seems_ to work.

Reply:
fwiw I figured out something less dumb than resolveOnTimeout: just await on the two promises we created